### DISCUSSION TITLE
How do you get annotation hints in your IDE?

### DISCUSSION THREAD

**Discussion by JorelAli** [Pante/Chimera#372] [General]
In your README, as well as in discussion #300, you've showcased an IDE with real-time compilation errors while you're editing the source code. I primarily use Eclipse for Java development and I understand that this may be an application you are unfamiliar with, but I have been trying for effectively a day to try and get real-time compilation errors working, so I can avoid having to stare at this in my IDE (using the typist-example-plugin):

![Commands class with a red underline](https://i.imgur.com/Y5KjB6M.png)

So far, I've tried following [various suggestions from StackOverflow](https://stackoverflow.com/questions/43404891/how-to-configure-java-annotation-processors-in-eclipse) which involved adding the annotation processor jar file to `Java Compiler > Annotation Processing > Factory Path`, but setting this appears to have no effect:

![](https://i.imgur.com/6adiotL.png)

I'm not entirely sure where to turn to now, so I thought I'd try my hand over here! _(Also, hi! 👋)_

**Comment by Pante** [comment]
Personally, I use NetBeans for Java development and it works out of the box without any additional configuration. I'm aware that IntelliJ does not support showing errors real-time, see [this issue](https://youtrack.jetbrains.com/issue/IDEA-134549?_ga=2.63096751.102128045.1649910100-1037338003.1562213966). However, it does detect the generated file. Nevertheless, I don't believe the lack of real-time errors to be the end of the world since it will still appear during `mvn install` or the gradle equivalent.

I haven't used Eclipse before, only NetBeans and IntelliJ so I'm rather unfamiliar with it myself 😅. Maybe you can try run `mvn install` on the project? That usually fixes any issues with detecting the file on NetBeans.

Also, hi 👋 !

**Comment by JorelAli** [comment]
> I'm aware that IntelliJ does not support showing errors real-time

This is very unexpected, given how powerful IntelliJ is!

-----

_(I am aware that this following information may mean very little to you, but I'm documenting it for the sake of anybody in the future that happens to come across this discussion with the exact same issue)_

After taking some inspiration of what you said (using `mvn install`), I discovered that there's this Eclipse plugin called [m2e-apt](https://github.com/jbosstools/m2e-apt) which claims to provide automatic annotation processing for Maven projects. On their GitHub page, the project is archived (in favour of the [Eclipse m2e project](https://github.com/eclipse-m2e/m2e-core/) which is installed in Eclipse by default). Apparently you have to tell Eclipse to update m2e which is a surprisingly unintuitive task to do. Instead, I just installed m2e-apt anyway and made more progress with that.

Running `mvn clean install` was very useful - as expected it generated the sources, but Eclipse _doesn't know about these by default_. I had to go to the project's `Properties > Maven > Annotation Processing` settings, and select the annotation processing mode for the maven project:

![Properties > Maven > Annotation Processing](https://i.imgur.com/XSrHyeo.png)

This tiny radio button basically sets up the rest of your maven project. It adds `target/generated-sources/annotations` (as well as test annotation sources) to the build path and it sets up the project's `Java Compiler > Annotation Processing > Factory Path` settings, populating it with the relevant annotation processors.

This fixes the missing `Commands.java` file! While I wish I could say this solved all of my problems, unfortunately Eclipse decides to complain about your annotation processor:

![Eclipse complaining about trying to load an annotation processor](https://i.imgur.com/z3W3y5J.png)

I highly doubt this is due to your annotation processor, and am happy to put the blame on "Eclipse is being weird" (I did a bit of testing with various other annotation processors and got the same error message). That said, thanks very much!

**Comment by JorelAli** [comment]
As a last desperate hope, I decided to update Eclipse (Install Eclipse 2022-03) and surprise surprise, everything _just works perfectly!_

![Working annotations in Eclipse!](https://i.imgur.com/D4hPWoW.png)

(This still required various setup in `Java Compiler > Annotation Processing` and `Java Compiler > Annotation Processing > Factory Path`)

**Comment by Pante** [comment]
Not at all, it's always nice to have someone document all these down, thanks!
